src/jit/morph.cpp

```diff
@@ -6940,263 +6945,236 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+                    //
+                    // Because of this requirement, we lose the opportunity to fast tail call
+                    // >8 byte size structs on non windows if either the callee or caller
+                    // have stack arguments.
+                     hasLargerThanOneStackSlotSizedStruct = true;
                 }
-
-#elif defined(WINDOWS_AMD64_ABI)
-
-                ++calleeArgRegCount;
-
-#endif // UNIX_AMD64_ABI
-
-#else
-                assert(!"Target platform ABI rules regarding passing struct type args in registers");
-                unreached();
-#endif //_TARGET_AMD64_ || _TARGET_ARM64_
             }
-            else
+
+            if (arg->passedByRef)
             {
-                hasMultiByteStackArgs = true;
+                hasByrefStruct = true;
+                break;
             }
         }
-        else
-        {
-            varTypeIsFloating(argx) ? ++calleeFloatArgRegCount : ++calleeArgRegCount;
-        }
+    }
+
+    if (callee->HasRetBufArg()) // RetBuf
+    {
+        // We don't increment callerIntArgRegCount here, since it is already in callee->gtCallArgs.
 
-        // We can break early on multiByte cases.
-        if (hasMultiByteStackArgs)
+        // If callee has RetBuf param, caller too must have it.
+        // Otherwise go the slow route.
+        if (info.compRetBuffArg == BAD_VAR_NUM)
         {
-            break;
+            reportFastTailCallDecision("Callee has RetBuf but caller does not.", 0, 0);
+            return false;
         }
     }
 
     const unsigned maxRegArgs = MAX_REG_ARG;
+    const size_t callerStackSize = info.compArgStackSize;
 
-// If we reached here means that callee has only those argument types which can be passed in
-// a register and if passed on stack will occupy exactly one stack slot in out-going arg area.
-// If we are passing args on stack for the callee and it has more args passed on stack than
-// the caller, then fast tail call cannot be performed.
-//
-// Note that the GC'ness of on stack args need not match since the arg setup area is marked
-// as non-interruptible for fast tail calls.
+    // Do not support byref struct types. The analysis required to prove that the struct will not
+    // create an outgoing local to pass to the callee is too much work.
+    //
+    // For example, consider the following case:
+    //
+    //  (48 byte struct)
+    //  {
+    //      a: {
+    //          long a;
+    //          long b;
+    //          long c;
+    //      },
+    //      b: {
+    //          long a;
+    //          long b;
+    //          long c;
+    //      }
+    //  }
+    //
+    // caller(48ByteStruct s)
+    // callee(s.a, s.b)
+    //
+    // This tail call is technically possible; however, it sets a later requirement
+    // that the outgoing args of s.a and s.b can only pass a pointer to the calleer's 
+    // fields. This information will be lost by the callee, and it will have similar
```

Comment:
Is this a platform-specific limitation? Would it necessarily be true for all platforms? Is there other code involved in the JIT in this restriction?

Reply:
I *think* what's being said here is that we can theoretically tail call in the situation where the caller has N bytes of byref-passed struct(s) that correspond with N bytes of callee byref-passed structs, but that it is difficult to ensure that all the structs are passed by reference. What I don't understand is why it matters that the information is lost by the callee. If the caller determines that the struct(s) is/are passed by reference to the callee, then the callee has that information. We just can't support the case (for example) where we have one incoming N byte struct that's passed by reference and the corresponding callee args are 2 N/2 byte structs, which are passed by value. Am I missing something?

Reply:
I wonder what the most general case is here. If the caller has `n` incoming by-ref struct arguments, of size s1, s2, ... sn, then do we have free reign to allocate those `n` independent buffers however we wish for the purpose of outgoing fast tailcall byref struct args? E.g., we could pass any number of by-ref args as long as we can pack them into the existing incoming `n` buffers.

Another question: what if the incoming by-ref structs have GC pointers? Does the incoming function's caller (aka, in the fast tailcall code, the "caller's caller") report the slots? Or does this function report them? If it's the caller's caller, then we can't use them for anything (except structs that match the GC types precisely).